Joss paper review (#51)
arfc/osier: paper/paper.md

---
title: '`Osier`: A Python package for multi-objective energy system optimization'
tags:
    - Python
    - energy systems
    - genetic algorithms
    - multi-objective optimization
authors:

>     - name: Samuel G. Dotson
>       orcid: 0000-0002-8662-0336
>       affiliation: 1 
>       corresponding: true
> affiliations:

Reviewer: ```suggestion
    - name: Samuel G. Dotson
      orcid: 0000-0002-8662-0336
      affiliation: 1 
      corresponding: true
    - name: Madicken Munk
      orcid: 0000-0003-0117-5366
      affiliation: 1 
affiliations:
```

>     - name: Felix T. Adler Fellow, Department of Nuclear, Plasma, and Radiological Engineering, University of Illinois Urbana-Champaign, USA

Reviewer: I don't believe your fellowship in an affiliation. It's part of your title. 

```suggestion
    - name: Department of Nuclear, Plasma, and Radiological Engineering, University of Illinois Urbana-Champaign, USA
```

      index: 1

> date: 20 February 2024

Reviewer: I think you should change this date to the date when you submit to JOSS.

bibliography: paper.bib
---

# Summary
Transitioning to a clean energy economy will require expanded energy
infrastructure. An equitable, or just, transition further requires the
recognition of the people and communities directly affected by this transition.
However, public preferences may be ignored during decision-making processes
related energy infrastructure due to a lack of technical rigor or expertise

> [@johnson:2021]. This challenge is further complicated by the fact that people

Reviewer: Is this the exact language from johnson? Would it be more appropriate to say this is a perceived lack of technical rigor or expertise? 

```suggestion
related to energy infrastructure due to a lack of technical rigor or expertise
[@johnson:2021]. This challenge is further complicated by the fact that people
```

have and express preferences over many dimensions simultaneously.
Multi-objective optimization offers a method to help decision makers and
stakeholders understand the problem and analyze tradeoffs among solutions
[@liebman:1976]. Although, to date, no multi-objective energy modelling

> frameworks exist. Open-source multi-objective energy system framework (`osier`)

Reviewer: I thought the one in France existed? Would it be more appropriate to say no open source multi-objective energy modelling frameworks exist?

is a Python package for designing and optimizing energy systems across an
arbitrary number of dimensions. `osier` was designed to help localized
communities articulate their energy preferences in a technical manner without
requiring extensive technical expertise. In order to facilitate more robust
tradeoff analysis, `osier` generates a set of  technology portfolios, called a

> Pareto front, with multi-objective optimization using evolutionary algorithms.

Reviewer: The ordering of this phrasing sounds like the technology portfolios are called a pareto front, not that the pareto front defines the technology portfolios. What about something like this? 

```suggestion
tradeoff analysis, `osier` generates a set of solutions, called a
Pareto front, that are composed of a number of technology portfolios. The Pareto front is calculated using multi-objective optimization using evolutionary algorithms.
```

`osier` also extends the common

> modelling-to-generate-alternatives (MGA) algorithm into many dimensions,

Reviewer: ```suggestion
modelling-to-generate-alternatives (MGA) algorithm into the dimensions of the solution space,
```

allowing users to investigate the near-optimal for appealing alternative

> solutions. In this way, `osier` may aid modelers in addressing procedural and

Reviewer: ```suggestion
solutions. This is novel in that most energy system modeling tools apply MGA to a single dimension. In this way, `osier` may aid modelers in addressing procedural and
```

> recognition justice.

Reviewer: Because the former is procedural, should this be recognitional?

# Statement of Need
There are myriad open- and closed-source energy system optimization models
(ESOMs) available [@pfenninger:2022]. ESOMs can be used for a variety of tasks
but are most frequently used for prescriptive analyses meant to guide

> decision-makers in planning processes. However, despite the many available
> models, all of these tools share a fundamental characteristic: Optimization over

Reviewer: ```suggestion
decision-makers in planning processes. However, despite the countless 
models available, all of these tools share a fundamental characteristic: Optimization over
```

Author: I think "countless" is a little too hyperbolic.

a single economic objective (e.g., total cost or social welfare).
Simultaneously, there is growing awareness of energy justice and calls for its
inclusion in energy models [@pfenninger:2014; @vagero:2023]. Some studies
attempted to incorporate local preferences into energy system design through

> multi-criteria decision analysis (MCDA) and community focus groups

Reviewer: What is the difference between MCDA and a multi objective energy modeling tool?

Author: MCDA focuses more on deciding among alternatives -- the way the alternatives are produced could be through MOO but they could also be generated through specific scenario analysis. At least, this is my understanding of the literature.

[@bertsch:2016; @mckenna:2018; @zelt:2019]. But these studies rely on tools with
pre-defined objectives which are difficult to modify. Without the ability to add
objectives that reflect the concerns of a community, the priorities of that
community will remain secondary to those of modellers and decision
makers. A flexible and extensible multi-objective framework that fulfills this
need has not yet been developed. The `osier` framework closes this gap.

# Design and Implementation

> In order to run `osier`, users are only required to supply an energy demand time

Reviewer: ```suggestion
In order to run `osier`, users are required to supply an energy demand time
```

> series. Users can optionally provide weather data to incorporate solar or wind

Reviewer: Is this the only thing they are required to provide? What about what technologies they want to use and the limitations of the model?

based on any quantifiable metric. The requirements for a bespoke objective are: 

1. The first argument must be a list of `osier.Technology` objects.
2. The second argument must be the results from an `osier.DispatchModel`. But
   this may be a simple placeholder with a default value of `None`.
3. The function must return a single numerical value.
4. The final requirement, is that all `osier.Technology` objects
   possess the attribute being optimized.

These two features acknowledge that a modeler cannot know *a priori* all
possible objectives or parameters of interest. Allowing users to define their
own objectives and modify technology objects (or simply build their own by
inheriting from the `osier.Technology` class) accounts for this limitation and
expands the potential for incorporating localized preferences. Lastly, in order
to account for unmodeled or unmodelable objectives, `osier` extends the
conventional MGA algorithm into N-dimensions by using a farthest-first-traversal
in the design space.

## Documentation

`osier` offers robust documentation with detailed usage examples at
[osier.readthedocs.io](https://osier.readthedocs.io).

# Acknowledgements

> Samuel Dotson was supported by the Nuclear Regulatory Commission Fellowship

Reviewer: Your acknowledgements should include the people who reviewed your PRs for Osier.

Reviewer: Use the credit taxonomy to define who did what for this research output: https://credit.niso.org/
https://theplosblog.plos.org/2022/01/contributorship/

> Samuel Dotson was supported by the Nuclear Regulatory Commission Fellowship

Reviewer: ```suggestion
Samuel Dotson, the corresponding and lead author of this publication is responsible 
for the conceptualization of osier, developing the osier as a software, in preparing 
this manuscript for publication, and for performing analysis to validate osier. 
Madicken Munk provided resources and supervision for the work, as well as assisted 
in the review and editing of the manuscript. Samuel Dotson was supported 
by the Nuclear Regulatory Commission Fellowship
```

program. This research was part of the Advanced Reactors and Fuel Cycles (ARFC)

> group in the Department of Nuclear, Plasma, and Radiological Engineering (NPRE)
> at the University of Illinois Urbana-Champaign. Additionally, Samuel Dotson was
> supported by the Felix T. Adler Fellowship through NPRE. The views and opinions
> expressed in this paper are those of the author alone.

Reviewer: ```suggestion
group in the Department of Nuclear, Plasma, and Radiological Engineering (NPRE)
at the University of 
Illinois Urbana-Champaign. To that end, the authors would like to acknowledge ARFC 
members Oleksander Yardas, Luke Seifert, Nathan Ryan, Amanda Bachmann, 
and Sun Myung Park for their contributions in reviewing pull requests 
supporting the creation of osier. 
Additionally, Samuel Dotson was
supported by the Felix T. Adler Fellowship through NPRE. 
Finally, the authors would like to thank the JOSS reviewers for their time 
and commentary in reviewing this manuscript. 
```

---
title: '`Osier`: A Python package for multi-objective energy system optimization'
tags:
    - Python
    - energy systems
    - genetic algorithms
    - multi-objective optimization
authors:
    - name: Samuel G. Dotson
      orcid: 0000-0002-8662-0336
      affiliation: 1 
      corresponding: true
    - name: Madicken Munk
      orcid: 0000-0003-0117-5366
      affiliation: 1 
affiliations:
    - name: Department of Nuclear, Plasma, and Radiological Engineering, University of Illinois Urbana-Champaign, USA
      index: 1
date: 05 April 2024
bibliography: paper.bib
---

# Summary
Transitioning to a clean energy economy will require expanded energy
infrastructure. An equitable, or just, transition further requires the
recognition of the people and communities directly affected by this transition.
However, public preferences may be ignored during decision-making processes
related to energy infrastructure due to a lack of technical rigor or expertise

> [@johnson:2021]. This challenge is further complicated by the fact that people

Reviewer: Is this the exact language from Johnson? Is it due to an actual lack of rigor or expertise, or a perceived lack of rigor or expertise? Would it be more appropriate to say a perceived lack of technical rigor or expertise?

I think this is important because this tool allows them to address the perception specifically, but doesn't make any assumptions about the expertise that they have or do not have.

Author: Here's a quote 

> While there may be other reasons the ICC never responded to Tripp’s allegations, her lack of expertise combined with the use of improper testimony helped ICC Staff sidestep the issue (Johson, 2021 p.7).

Reviewer: Ok, then we can leave it. I do think for your future publications we should consider using language around osier and your tooling that clarifies its use in targeting a perception of lack of technical rigor.

Simultaneously, there is growing awareness of energy justice and calls for its
inclusion in energy models [@pfenninger:2014; @vagero:2023]. Some studies
incorporate local preferences into energy system design through
multi-criteria decision analysis (MCDA) and community focus groups
[@bertsch:2016; @mckenna:2018; @zelt:2019]. But these studies rely on tools with
pre-defined objectives which are difficult to modify. Without the ability to add
objectives that reflect the concerns of a community, the priorities of that
community will remain secondary to those of modellers and decision makers. A
flexible and extensible multi-objective framework that fulfills this need has
not yet been developed. The `osier` framework closes this gap.

# Design and Implementation
The fundamental object in `osier` is an `osier.Technology` object, which 
contains all of the necessary cost and performance data for different technology 
classes. `osier` comes pre-loaded with a variety of technologies described in 
the National Renewable Energy Laboratory's (NREL) Annual Technology Baseline (ATB)
dataset[@nationalrenewableenergylaboratory:2023] but users are also able to

> define their own. In order to run `osier`, users are only required to supply an 

Reviewer: For some reason this was missed in inclusion in the last round. `only` is not adding to your messaging here and makes your language less formal. 

```suggestion
define their own. In order to run `osier`, users are required to supply an 
```

system models, `osier` has no required objectives. While users may choose from a
variety of pre-defined objectives, they may also declare their own objectives
based on any quantifiable metric. The requirements for a bespoke objective are: 

1. The first argument must be a list of `osier.Technology` objects.
2. The second argument must be the results from an `osier.DispatchModel`. But
   this may be a simple placeholder with a default value of `None`.
3. The function must return a single numerical value.
4. The final requirement, is that all `osier.Technology` objects possess the
   attribute being optimized.

These two features acknowledge that a modeler cannot know *a priori* all
possible objectives or parameters of interest. Allowing users to define their
own objectives and modify technology objects (or simply build their own by
inheriting from the `osier.Technology` class) accounts for this limitation and
expands the potential for incorporating localized preferences. Lastly, in order
to account for unmodeled or unmodelable objectives, `osier` extends the
conventional MGA algorithm into N-dimensions by using a farthest-first-traversal

> in the design space.

Reviewer: This is a good rephrase!

## Sample Results and Interpretation

When solving a multi-objective problem, `osier` generates a set of co-optimal
solutions rather than a global optimum, called a Pareto front.
\autoref{fig:osier-results} shows a Pareto front from a problem that
simultaneously minimizes total cost and lifecylce carbon emissions.

![A Pareto front generated by`osier`.\label{fig:osier-results}](images/osier-results.png)

Each point on this Pareto front represents a different technology portfolio
(i.e., different combination of wind, natural gas, and battery storage).
\autoref{fig:osier-tech-res} illustrates the variation in solutions from 
the Pareto front in \autoref{fig:osier-results}. In this case, the range of 
wind capacity is wider than the range of capacities for natural gas and battery 
storage.

> ![The variance in technology options along a Pareto front.\label{fig:osier-tech-res}](images/osier-tech-results.png)

Reviewer: These added figures are good illustrations/additions. Very nice.
